From e42d4fc90bfd823e00a78731d899e2e35f2f732b Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Tue, 31 Oct 2017 10:50:14 +0800
Subject: [PATCH 2738/5242] MLK-16727-3 usb: cdns3: force vbus status
 accordingly

commit  f78429aa5b9f3547780cd09f161454941083fd90 from
https://source.codeaurora.org/external/imx/linux-imx.git

Since the controller doesn't know vbus status well due to IC limitation,
it needs to force vbus status for controller when the connection and
disconnection occur.

BuildInfo:
- SCFW 8dcff26, IMX-MKIMAGE ea027c4b, ATF
- U-Boot 2017.03-imx_v2017.03_4.9.51_imx8_beta1+g6dc7b0f

Acked-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/core.c |   11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/cdns3/core.c b/drivers/usb/cdns3/core.c
index 3524909..66264e0 100644
--- a/drivers/usb/cdns3/core.c
+++ b/drivers/usb/cdns3/core.c
@@ -117,9 +117,6 @@ static void cdns3_usb_phy_init(void __iomem *regs)
 	writel(0x01e0, regs + TB_ADDR_TX_RCVDET_ST_TMR);
 	writel(0x0090, regs + TB_ADDR_XCVR_DIAG_LANE_FCM_EN_MGN_TMR);
 
-	/* Force B Session Valid as 1 */
-	writel(0x0060, regs + 0x380a4);
-
 	udelay(10);
 
 	pr_debug("end of %s\n", __func__);
@@ -384,10 +381,16 @@ static int cdsn3_do_role_switch(struct cdns3 *cdns, enum cdns3_roles role)
 
 	current_role = cdns->role;
 	cdns3_role_stop(cdns);
-	if (role == CDNS3_ROLE_END)
+	if (role == CDNS3_ROLE_END) {
+		/* Force B Session Valid as 0 */
+		writel(0x0040, cdns->phy_regs + 0x380a4);
 		return 0;
+	}
 
 	cdns_set_role(cdns, role);
+	if (role == CDNS3_ROLE_GADGET || role == CDNS3_ROLE_HOST)
+		/* Force B Session Valid as 1 */
+		writel(0x0060, cdns->phy_regs + 0x380a4);
 	ret = cdns3_role_start(cdns, role);
 	if (ret) {
 		/* Back to current role */
-- 
1.7.9.5

